File: LeetCode/Medium/LRU Cache/Java/lruCache.java
class ListNode {
    int key;
    int val;
    ListNode next;
    ListNode prev;
    public ListNode(int key, int val) {
        this.key = key;
        this.val = val;
    }
}

class LRUCache {
    int cap;
    Map<Integer, ListNode> map;
    ListNode head;
    ListNode tail;
    public LRUCache(int capacity) {
        cap = capacity;
        map = new HashMap<>();
        head = new ListNode(-1, -1);
        tail = new ListNode(-1, -1);
        head.next = tail;
        tail.prev = head;
    }
    
    public int get(int key) {
        if (!map.containsKey(key)) return -1;
        ListNode node = map.get(key);
        remove(node);
        add(node);
        return node.val;
    }
    
    public void put(int key, int value) {
        if (map.containsKey(key)) {
            ListNode oldNode = map.get(key);
            remove(oldNode);
        }

        ListNode node = new ListNode(key, value);
        map.put(key, node);
        add(node);
        
        if (map.size() > cap) {
            ListNode del = head.next;
            remove(del);
            map.remove(del.key);
        }
    }

    private void add(ListNode node) {
        ListNode prev = tail.prev;
        prev.next = node;
        node.prev = prev;
        node.next = tail;
        tail.prev = node;
    }

    private void remove(ListNode node) {
        node.prev.next = node.next;
        node.next.prev = node.prev;
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */    

//a better approach
class LRUCache extends LinkedHashMap<Integer, Integer> {
    private final int capacity;

    public LRUCache(int capacity) {
        super(capacity, 0.75f, true);
        this.capacity = capacity;    
    }
    
    public int get(int key) {
        return super.getOrDefault(key, -1);
    }
    
    public void put(int key, int value) {
        super.put(key, value);
    }
    
    @Override
    protected boolean removeEldestEntry(Map.Entry<Integer, Integer> eldest) {
        return size() > capacity;
    }
}

File: Algorithms/Coding Interview Patterns/Linked List/lru_cache.cc
#include <iostream>
#include <unordered_map>
#include "doubly_ll.h"

using namespace std;

class LRU {
    unordered_map<int, Node*> hash_map;
    int cap;
    Node* head;
    Node* tail;

public:
    LRU(int cap) {
        this->cap = cap;
        this->head = new Node;
        this->tail = new Node;
        this->head->key = -1;
        this->head->val = -1;
        this->tail->key = -1;
        this->tail->val = -1;
        this->head->next = this->tail;
        this->tail->prev = this->head;
    }

    int get(int key) {
        if (this->hash_map.find(key) == this->hash_map.end()) return -1;
        
        remove(this->hash_map[key]);
        add_to_tail(this->hash_map[key]);
        return this->hash_map[key]->val;
    }

    void put(int key, int val) {
        if (this->hash_map.find(key) != this->hash_map.end()) remove(this->hash_map[key]);
        Node* node = new Node;
        node->key = key;
        node->val = val;
        this->hash_map[key] = node;

        if (this->hash_map.size() > this->cap) {
            this->hash_map.erase(this->head->next->key);
            remove(this->head->next);
        }
        add_to_tail(node);
    } 

    void add_to_tail(Node* node) {
        Node* prev = this->tail->prev;
        node->prev = prev;
        node->next = this->tail;
        prev->next = node;
        this->tail->prev = node;
    }

    void remove(Node* node) {
        node->prev->next = node->next;
        node->next->prev = node->prev;
    }
};

int main(int argc, char **argv) {
    LRU lru(3);

    lru.put(2, 250);
    lru.put(4, 300);
    lru.put(3, 200);
    cout << lru.get(2) << "\n";
    cout << lru.get(3) << "\n";
    return 0;
}

